refactor(tests): integrate lua-client into core and use core for functionaltests

bfredl opened the pull request:
more exciting errors for actually running more code under the sanitizers!

bfredl committed:
refactor(tests): move lua-client into core and use it for functionaltests

Eliminates lua-client and non-static libluv as test time dependencies

Note: the API for a public lua-client is not yet finished.
The interface needs to be adjusted to work in the embedded loop
of a nvim instance (to use it to talk between instances)

bfredl committed:
fix(tests): fixes for using vim.mpack and more ASAN

justinmk commented on cmake/RunTests.cmake:
```diff
@@ -71,16 +71,11 @@ if(NOT DEFINED ENV{TEST_TIMEOUT} OR "$ENV{TEST_TIMEOUT}" STREQUAL "")
 
 execute_process(
-  COMMAND ${RUNNER_PRG} -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
+  COMMAND ${NVIM_PRG} -ll ${WORKING_DIR}/test/busted_runner.lua -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
```
```suggestion
  COMMAND ${NVIM_PRG} -l ${WORKING_DIR}/test/busted_runner.lua -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
```

justinmk commented on cmake/RunTests.cmake:
```diff
@@ -71,16 +71,11 @@ if(NOT DEFINED ENV{TEST_TIMEOUT} OR "$ENV{TEST_TIMEOUT}" STREQUAL "")
 
 execute_process(
-  COMMAND ${RUNNER_PRG} -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
+  COMMAND ${NVIM_PRG} -ll ${WORKING_DIR}/test/busted_runner.lua -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
```
i see `-ll` was used in the old code but should it be `-l` ?

bfredl replied:
should be `-ll` as is. we want full control of initialization in tests.

justinmk commented on test/client/session.lua:
```diff
@@ -0,0 +1,194 @@
+local safe_pcall
```
is `test/client/session.lua` actually only for tests, or (eventually) would we move it to `runtime/...` ?

bfredl replied:
eventually. regardless, `coxpcall` is very small and pure lua, so we could vendor it in `runtime/` when this happens.

zeertzjq commented on test/functional/helpers.lua:
```diff
@@ -1,14 +1,12 @@
-local TcpStream = require('nvim.tcp_stream')
-local SocketStream = require('nvim.socket_stream')
-local ChildProcessStream = require('nvim.child_process_stream')
+-- nvim client: Found in runtime/vim/client :)
```
```suggestion
-- nvim client: Found in test/client :)
```

zeertzjq commented on test/client/uv_stream.lua:
```diff
@@ -0,0 +1,164 @@
+    self._proc:kill('sig'..signal)
+  end
+  uv.run('nowait')
+end
```
Does this no longer wait for the process to exit? Then some warnings about session taking too long to exit in the test suite may no longer make sense.

zeertzjq replied:
It's possible to use `expect_exit(command, 'qall!')` to make sure the process exits, but that may be a bit slower as it makes an RPC request.

bfredl replied:
we could add a wait for exit the proper way (run loop until exit callback, with a timeout)

bfredl committed:
fix(client): wait for session to exit

This replicates the old native.pid_wait(self._pid)
call, except using the proper libuv pattern (run loop unitil exit
callback)
